### PR TITLE
Fix connection leakage in CustomRestClient

### LINKED ISSUE #1471: Connection leakage in CustomRestClient
### Describe the bug
We found that `ConnectionPoolTimeoutException` is thrown from `CustomRestClientImpl` and cause custom health check failure.

```
[INFO ] 2020-10-16 13:10:54,273 org.apache.helix.rest.client.CustomRestClient - Executing request: POST http://localhost:11933/instanceHealthStatus HTTP/1.1, headers:
[Accept: application/json], entity: [Content-Type: application/json; charset=UTF-8,Content-Length: 2,Chunked: false]
[ERROR] 2020-10-16 13:11:54,281 org.apache.helix.rest.client.CustomRestClient - Failed to perform customized health check. Is participant endpoint http://localhost:119
33/instanceHealthStatus available?
org.apache.http.conn.ConnectionPoolTimeoutException: Timeout waiting for connection from pool
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.leaseConnection(PoolingHttpClientConnectionManager.java:314)
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager$1.get(PoolingHttpClientConnectionManager.java:280)
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:190)
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
        at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
        at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
        at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56)
        at org.apache.helix.rest.client.CustomRestClientImpl.post(CustomRestClientImpl.java:140)
        at org.apache.helix.rest.client.CustomRestClientImpl.getInstanceStoppableCheck(CustomRestClientImpl.java:75)
        at org.apache.helix.rest.server.service.InstanceServiceImpl.performCustomInstanceCheck(InstanceServiceImpl.java:259)
        at org.apache.helix.rest.server.service.InstanceServiceImpl.lambda$null$2(InstanceServiceImpl.java:200)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

### To Reproduce
Change code a bit to let `handleResponse` always go to the else branch that does not consume entity and throw `ClientProtocolException`.

### Expected behavior
Should not have `ConnectionPoolTimeoutException`.

### Additional context
If custom health check rest calls return non-200 responses, the entity is not consumed so the connection is still occupied, which causes connection leakage. We need to consume the response entity before throwing an exception for non-200 responses, so the stream or connection is released.

### PR DESCRIPTION
**Pull request by pkuwm**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Fixes #1471 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

There is connection leakage in `CustomRestClientImpl` and causes timeout waiting for connection.

Fix this issue by consuming entity and releasing stream and connection.

### Tests

- [x] The following tests are written for this issue:

 Manually reproduce the issue locally. After the fix, the issue is gone.

- [x] The following is the result of the "mvn test" command on the appropriate module:

```
[INFO] Tests run: 170, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 100.208 s - in TestSuite
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 170, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:46 min
[INFO] Finished at: 2020-10-16T15:20:24-07:00
[INFO] ------------------------------------------------------------------------
```

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java [hunk 1]
```diff
@@ -40,7 +40,7 @@
 import org.slf4j.LoggerFactory;
 
 class CustomRestClientImpl implements CustomRestClient {
-  private static final Logger LOG = LoggerFactory.getLogger(CustomRestClient.class);
```
Comment (kaisun2000) [inline helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java:43]: Good catch.

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java [hunk 2]
```diff
@@ -116,25 +116,36 @@ protected JsonNode getJsonObject(HttpResponse httpResponse) throws IOException {
...
-      HttpPost postRequest = new HttpPost(url);
       postRequest.setHeader("Accept", ACCEPT_CONTENT_TYPE);
       StringEntity entity = new StringEntity(OBJECT_MAPPER.writeValueAsString(payloads),
           ContentType.APPLICATION_JSON);
       postRequest.setEntity(entity);
       LOG.info("Executing request: {}, headers: {}, entity: {}", postRequest.getRequestLine(),
           postRequest.getAllHeaders(), postRequest.getEntity());
-      return _httpClient.execute(postRequest);
+
+      HttpResponse response = _httpClient.execute(postRequest);
```
Comment (kaisun2000) [inline helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java:137]: handleResponse() handles the result from post() assuming post has not IO exception. That is the original author's thinking.  So looking at the response code previously is in handleResponse(). If sticking to the same line of thinking, we can add this warning log to handleResponse, instead in post() here.

But in reality, it does not really matter. More like a style of personal choice.

**Review by kaisun2000 — APPROVED** [review]

**Comment by kaisun2000** [comment]
@pkuwm, yes, this should fix the issue I found which can be urgent. So without a unit test, it is fine for now.

Can you also add an issue to add a unit test?

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java (hunk 2 above)
Comment (pkuwm) [reply helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java:137]: I tried that. If we move to `handleResponse`, we'll have to change more code, eg. method signature of `handleResponse` to have an extra parameter `payloads`, and we need to change the callers as well. That sounds more work to me and it doesn't look better than this. I prefer this way so it is cleaner.

**Comment by pkuwm** [comment]
> @pkuwm, yes, this should fix the issue I found which can be urgent. So without a unit test, it is fine for now.
> 
> Can you also add an issue to add a unit test?

I actually thought about unit testing the connection leak and did an investigation on it. It is not easy. As long as we use it correctly, it is fine.
One way is, once we switch to use an instrumented http client, we can check the metrics in the unit test to test the connection leak. 
Created a ticket: https://github.com/apache/helix/issues/1475 for tracking.

**Comment by pkuwm** [comment]
This PR is ready to be merged, approved by @kaisun2000 

There is connection leakage in CustomRestClientImpl and causes timeout waiting for connection.
Fix this issue by consuming entity and releasing the content stream and connection.

**Pull request merged**